components/service/glean/src/main/java/mozilla/components/service/glean/Glean.kt

```diff
@@ -154,6 +172,56 @@ open class GleanInternalAPI internal constructor () {
         return uploadEnabled
     }
 
+    /**
+     * Handles the changing of state when uploadEnabled changes.
+     *
+     * When disabling, all pending metrics, events and queued pings are cleared.
+     *
+     * When enabling, the core Glean metrics are recreated.
+     *
+     * If Glean is not initialized, this is a no-op.
+     */
+    private fun onChangeUploadEnabled(enabled: Boolean) {
+        if (enabled) {
+            initializeCoreMetrics(applicationContext!!)
+        } else {
+            clearMetrics()
+        }
+    }
+
+    /**
+     * Clear any pending metrics when telemetry is disabled.
+     */
+    @Suppress("EXPERIMENTAL_API_USAGE")
+    private fun clearMetrics() = Dispatchers.API.launch {
```

Comment:
Personally I find this style of "inlining" the Dispatcher call a bit weird.
IMO it hides that it is async.

Reply:
Sure, it takes some getting used to.  I'd suggest moving that into its own bug, though, since this pattern is used in a number of places already.

Reply:
Ah yeah, no need to handle it here.